src/ICSharpCode.SharpZipLib/Tar/TarOutputStream.cs

```diff
@@ -252,23 +287,23 @@ public void PutNextEntry(TarEntry entry)
 				longHeader.GroupName = entry.GroupName;
 				longHeader.UserName = entry.UserName;
 				longHeader.LinkName = "";
-				longHeader.Size = entry.TarHeader.Name.Length + 1;  // Plus one to avoid dropping last char
+				longHeader.Size = namelen + 1;  // Plus one to avoid dropping last char
 
-				longHeader.WriteHeader(blockBuffer);
+				longHeader.WriteHeader(blockBuffer, encoding);
 				buffer.WriteBlock(blockBuffer);  // Add special long filename header block
 
 				int nameCharIndex = 0;
 
-				while (nameCharIndex < entry.TarHeader.Name.Length + 1 /* we've allocated one for the null char, now we must make sure it gets written out */)
+				while (nameCharIndex < namelen + 1 /* we've allocated one for the null char, now we must make sure it gets written out */)
 				{
 					Array.Clear(blockBuffer, 0, blockBuffer.Length);
-					TarHeader.GetAsciiBytes(entry.TarHeader.Name, nameCharIndex, this.blockBuffer, 0, TarBuffer.BlockSize); // This func handles OK the extra char out of string length
+					TarHeader.GetAsciiBytes(entry.TarHeader.Name, nameCharIndex, this.blockBuffer, 0, TarBuffer.BlockSize, encoding); // This func handles OK the extra char out of string length
```

Comment:
This should probably be renamed?

Reply:
Indeed, it should be renamed to `nameEncoding`

Reply:
Sorry, I meant the "GetAsciiBytes" method. The name is not really applicable when another encoding is used.